### PR TITLE
refactor to merge replayer and berkeley_account_tables app

### PR DESCRIPTION
**Pull request by ghost-not-in-the-shell**
Explain your changes:
This PR did a refactor on `replayer` to add the functionality of `berkeley_account_tables` app. Since the `replayer` and `berkeley_account_tables` are almost same, so I did such a refactor to `replayer` so that if you pass in the `--migration-mode` flag, then it would function as the `berkeley_account_tables` app.

Explain how you tested your changes:
*

Checklist:

- [ ] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

* Closes #0000

### CODE REVIEW

**Review by dkijania — APPROVED** [review]

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

File: src/app/replayer/replayer.ml [hunk 1]
```diff
@@ -1311,8 +1313,61 @@ let main ~input_file ~output_file_opt ~archive_uri ~continue_on_error
                     ] ;
                 Deferred.unit )
               else
+                let%bind accounts_before = Ledger.to_list ledger in
```
Comment (georgeee) [inline src/app/replayer/replayer.ml:1316]: I this an expensive operation? May it make sense to perform the added operations only in migration mode (as the results os this seem to be discarded in non-migration mode)

**Review by georgeee — APPROVED** [review]

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Review by amc-ie — APPROVED** [review]

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by deepthiskumar** [comment]
What is the snarkyJS submodule update for?

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
> What is the snarkyJS submodule update for?

It's caused by this commit https://github.com/MinaProtocol/mina/pull/14720/commits/f8b2bff319cf849de5efe654abe0ba4c894aa669

I forgot to update the snarkyjs submodule when pulling from berkeley

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

File: src/app/replayer/replayer.ml [hunk 2]
```diff
@@ -1601,6 +1660,14 @@ let () =
            Param.flag "--output-file"
              ~doc:"file File containing the resulting ledger"
              Param.(optional string)
+         and migration_mode =
+           Param.flag "--migration-mode"
+             ~doc:
+               "If this flag is turned on then migration mode would be \
+                started, which means ledger hash check would be disabled and \
+                this app would populates the `accounts_accessed` and \
+                `accounts_created` tables"
```
Comment (deepthiskumar) [inline src/app/replayer/replayer.ml:1669]: Could you make this concise? Maybe something like-
```
Disable ledger hash check and populate accounts_accessed table. Used only for migrating mainnet database for berkeley upgrade
```

File: src/app/replayer/replayer.ml [hunk 3]
```diff
@@ -1601,6 +1660,14 @@ let () =
            Param.flag "--output-file"
              ~doc:"file File containing the resulting ledger"
              Param.(optional string)
+         and migration_mode =
+           Param.flag "--migration-mode"
```
Comment (deepthiskumar) [inline src/app/replayer/replayer.ml:1664]: Can we rename this to `berkeley-migration-mode`? it is very specific to berkeley upgrade

**Review by deepthiskumar — DISMISSED** [review]
Ignoring ledger hash is probably the case for all HFs but populating `accounts_accessed` table is very berkeley specific which will be used only till the migration is completed.  So it seems a bit weird to have this in the replayer code that will be maintained till the next HF. Is the plan to remove the migration mode after the upgrade? 
Also, it is not clear to me how this will be used to check previous fork's data, new data from the upgrade (especially around the fork point where the replayer has to check both previous and new data). I remember we discussed about specifying a global slot instead. For blocks before the slot, the replayer will ignore hashes (and it could populate `accounts_accessed` table) and for the slot after it should check hashes.

**Comment by ghost-not-in-the-shell** [comment]
> Ignoring ledger hash is probably the case for all HFs but populating `accounts_accessed` table is very berkeley specific which will be used only till the migration is completed. So it seems a bit weird to have this in the replayer code that will be maintained till the next HF. Is the plan to remove the migration mode after the upgrade? Also, it is not clear to me how this will be used to check previous fork's data, new data from the upgrade (especially around the fork point where the replayer has to check both previous and new data). I remember we discussed about specifying a global slot instead. For blocks before the slot, the replayer will ignore hashes (and it could populate `accounts_accessed` table) and for the slot after it should check hashes.

This PR did what it says. Since the `berkeley_account_tables` is almost identical to the `replayer`, so I created this PR to merge these 2. It didn't do what you says. I can create another PR that do the check for 1. previous fork's data, 2. new data from the upgrade.

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!approved-for-mainnet

**Pull request merged**